lib/internal/dns.js

```diff
@@ -0,0 +1,2624 @@
+      main resolver to avoid creating a new Buffer every time we need to retry
+      the same query on a different protocol or nameserver
+    - Add file (and registry, on Windows) watchers for updates to system DNS
+      settings
+      - If this is implemented, how should this interact with settings
+        overridden by users (e.g. via setServers())?
+    - Support EDNS, DNSSEC
+    - Allow complete override of system-level DNS options in resolve() instead
+      (or in addition to?) being able to append to existing options
+      (e.g. system has `use_vc` set, but user wants to un-set that for a
+      particular resolve())
+
+  NOTES:
+    - PR #744 mentioned c-ares could return *mixed* IPv4 and IPv6 when
+      `family === 0`, but this is not true (it's even stated in c-ares' TODO),
+      so this implementation does not return mixed results either
+*/
+'use strict';
+
+const TCPSocket = require('net').Socket;
+const UDPSocket = require('dgram').Socket;
+const isIP = require('net').isIP; // TODO: move implementation from cares_wrap
+const cp = require('child_process');
+const os = require('os');
+const fs = require('fs');
+const pathJoin = require('path').join;
+
+const HOSTNAME = os.hostname();
+const REGEX_ARRAY = /^\{?(.*)\}?$/;
+const REGEX_OPTS =
+    /^(ndots|timeout|attempts|rotate|inet6|use-vc)(?::(\d{1,2}))?$/;
+const REGEX_TYPES = /^(nameserver|domain|search|sortlist|options|lookup) /;
+const REGEX_WHITESPACE = /[ \f\t\v]+/g;
+const REGEX_COMMENT = /[#;].*$/;
+const LOOKUPNS = ['dns'];
+const QTYPE_A = 1;
+const QTYPE_PTR = 12;
+const QTYPE_AAAA = 28;
+const FLAG_HAS_IPv4 = 0x1;
+const FLAG_HAS_IPv6 = 0x2;
+const FLAG_HAS_IPv4_IPv6 = (FLAG_HAS_IPv4 | FLAG_HAS_IPv6);
+const V4MAPPED = 0x08;
+const ADDRCONFIG = 0x20;
+const FLAG_RESOLVE_NSONLY = 0x01; // No local hosts file lookup
+const FLAG_RESOLVE_NOSEARCH = 0x02; // No domain searching
+const FLAG_RESOLVE_NOALIASES = 0x04; // No HOSTALIASES checking
+const FLAG_RESOLVE_IGNTC = 0x08; // Ignore truncation errors (UDP)
+const FLAG_RESOLVE_USEVC = 0x10; // Always use TCP a.k.a "Virtual Circuit"
+const DNS_DEFAULTS = {
+  // TODO: improve default nameserver for IPv6-only systems
+  nameserver: [ '127.0.0.1' ],
+  search: (~HOSTNAME.indexOf('.')
+           ? [ HOSTNAME.slice(HOSTNAME.indexOf('.') + 1) ] : null),
+  sortlist: null,
+  options: {
+    ndots: 1,
+    timeout: 5000, // ms
+    attempts: 4, // how many tries before giving up
+    rotate: false, // round-robin nameservers instead of always trying first one
+    inet6: false, // query AAAA before A
+    use_vc: false // force TCP for all DNS queries
+  },
+  lookup: [ 'files', 'dns' ]
+};
+
+// Map DNS RCODE values to sensible string values
+const RCODE_V2S = {
+  // RFC 1035
+  0: 'NOERROR', // No error
+  1: 'FORMERR', // Format error
+  2: 'SERVFAIL', // Server failure
+  3: 'NXDOMAIN', // Non-existent domain
+  4: 'NOTIMP', // Not implemented
+  5: 'REFUSED', // Query refused
```

Comment:
Would it make sense to add RCODEs 6 and above? I have no experience with them, but if they can happen, it might be best to have them. Also, trailing comma (not that it matters).